File: platforms/tauri/src/shortcuts.rs
use tauri::{AppHandle, Manager, Runtime};
use tauri_plugin_global_shortcut::{GlobalShortcutExt, Shortcut};

pub fn setup_shortcuts<R: Runtime>(app: &AppHandle<R>) -> Result<(), Box<dyn std::error::Error>> {
    // Register global shortcuts

    // Toggle window visibility: Cmd/Ctrl+Shift+Space
    let toggle_window_shortcut = "CmdOrCtrl+Shift+Space";
    app.global_shortcut().on_shortcut(toggle_window_shortcut, {
        let app_handle = app.clone();
        move || {
            log::debug!("Global shortcut triggered: toggle window");
            if let Some(window) = app_handle.get_webview_window("main") {
                if window.is_visible().unwrap_or(false) {
                    let _ = window.hide();
                } else {
                    let _ = window.show();
                    let _ = window.set_focus();
                }
            }
        }
    })?;

    // Show window: Cmd/Ctrl+Shift+N
    let show_window_shortcut = "CmdOrCtrl+Shift+N";
    app.global_shortcut().on_shortcut(show_window_shortcut, {
        let app_handle = app.clone();
        move || {
            log::debug!("Global shortcut triggered: show window");
            if let Some(window) = app_handle.get_webview_window("main") {
                let _ = window.show();
                let _ = window.set_focus();
            }
        }
    })?;

    // Quick voice call toggle: Cmd/Ctrl+Shift+V
    let voice_call_shortcut = "CmdOrCtrl+Shift+V";
    app.global_shortcut().on_shortcut(voice_call_shortcut, {
        let app_handle = app.clone();
        move || {
            log::debug!("Global shortcut triggered: voice call");
            if let Some(window) = app_handle.get_webview_window("main") {
                let _ = window.emit("shortcut-voice-call", ());
            }
        }
    })?;

    // Mute/unmute toggle: Cmd/Ctrl+Shift+M
    let mute_shortcut = "CmdOrCtrl+Shift+M";
    app.global_shortcut().on_shortcut(mute_shortcut, {
        let app_handle = app.clone();
        move || {
            log::debug!("Global shortcut triggered: mute toggle");
            if let Some(window) = app_handle.get_webview_window("main") {
                let _ = window.emit("shortcut-mute-toggle", ());
            }
        }
    })?;

    log::info!("Global shortcuts registered successfully");
    Ok(())
}

/// Register a custom global shortcut
#[tauri::command]
pub async fn register_shortcut<R: Runtime>(
    app: AppHandle<R>,
    shortcut: String,
    action: String,
) -> Result<(), String> {
    app.global_shortcut()
        .on_shortcut(&shortcut, move || {
            log::debug!("Custom shortcut triggered: {} -> {}", shortcut, action);
        })
        .map_err(|e| e.to_string())
}

/// Unregister a global shortcut
#[tauri::command]
pub async fn unregister_shortcut<R: Runtime>(
    app: AppHandle<R>,
    shortcut: String,
) -> Result<(), String> {
    app.global_shortcut()
        .unregister(&shortcut)
        .map_err(|e| e.to_string())
}

/// Check if a shortcut is registered
#[tauri::command]
pub async fn is_shortcut_registered<R: Runtime>(
    app: AppHandle<R>,
    shortcut: String,
) -> Result<bool, String> {
    app.global_shortcut()
        .is_registered(&shortcut)
        .map_err(|e| e.to_string())
}

File: platforms/electron/main/shortcuts.ts
/**
 * Global Keyboard Shortcuts
 *
 * Registers global keyboard shortcuts that work even when the app is not focused.
 * Useful for quick access, voice calls, and other quick actions.
 */

import { app, globalShortcut } from 'electron'
import log from 'electron-log'
import { showMainWindow, toggleMainWindow, getMainWindow } from './window'
import settingsStore from './store'

export function registerGlobalShortcuts(): void {
  try {
    // Get custom shortcuts from settings or use defaults
    const shortcuts = settingsStore.get('globalShortcuts') || {
      toggleWindow: 'CommandOrControl+Shift+Space',
      showWindow: 'CommandOrControl+Shift+N',
      voiceCall: 'CommandOrControl+Shift+V',
      muteToggle: 'CommandOrControl+Shift+M',
    }

    // Toggle window visibility
    if (shortcuts.toggleWindow) {
      const registered = globalShortcut.register(shortcuts.toggleWindow, () => {
        log.debug('Global shortcut: Toggle window')
        toggleMainWindow()
      })
      if (registered) {
        log.info(`Global shortcut registered: ${shortcuts.toggleWindow} (Toggle window)`)
      } else {
        log.warn(`Failed to register global shortcut: ${shortcuts.toggleWindow}`)
      }
    }

    // Show window (always brings to front)
    if (shortcuts.showWindow) {
      const registered = globalShortcut.register(shortcuts.showWindow, () => {
        log.debug('Global shortcut: Show window')
        showMainWindow()
      })
      if (registered) {
        log.info(`Global shortcut registered: ${shortcuts.showWindow} (Show window)`)
      } else {
        log.warn(`Failed to register global shortcut: ${shortcuts.showWindow}`)
      }
    }

    // Quick voice call toggle
    if (shortcuts.voiceCall) {
      const registered = globalShortcut.register(shortcuts.voiceCall, () => {
        log.debug('Global shortcut: Voice call')
        const mainWindow = getMainWindow()
        mainWindow?.webContents.send('app:toggle-voice-call')
      })
      if (registered) {
        log.info(`Global shortcut registered: ${shortcuts.voiceCall} (Voice call)`)
      } else {
        log.warn(`Failed to register global shortcut: ${shortcuts.voiceCall}`)
      }
    }

    // Mute/unmute toggle
    if (shortcuts.muteToggle) {
      const registered = globalShortcut.register(shortcuts.muteToggle, () => {
        log.debug('Global shortcut: Mute toggle')
        const mainWindow = getMainWindow()
        mainWindow?.webContents.send('app:toggle-mute')
      })
      if (registered) {
        log.info(`Global shortcut registered: ${shortcuts.muteToggle} (Mute toggle)`)
      } else {
        log.warn(`Failed to register global shortcut: ${shortcuts.muteToggle}`)
      }
    }

    log.info('Global shortcuts registered successfully')
  } catch (error) {
    log.error('Failed to register global shortcuts:', error)
  }
}

export function unregisterGlobalShortcuts(): void {
  globalShortcut.unregisterAll()
  log.info('All global shortcuts unregistered')
}

// Re-register shortcuts when settings change
export function updateGlobalShortcuts(): void {
  unregisterGlobalShortcuts()
  registerGlobalShortcuts()
  log.info('Global shortcuts updated')
}

// Clean up shortcuts when app quits
app.on('will-quit', () => {
  unregisterGlobalShortcuts()
})

// Check if a shortcut is registered
export function isShortcutRegistered(accelerator: string): boolean {
  return globalShortcut.isRegistered(accelerator)
}

// Get all registered shortcuts
export function getRegisteredShortcuts(): string[] {
  const shortcuts = settingsStore.get('globalShortcuts') || {}
  return Object.values(shortcuts).filter(Boolean) as string[]
}
